velox/functions/lib/Re2Functions.cpp

```diff
@@ -557,55 +583,114 @@ std::pair<bool, int32_t> matchRelaxedFixedForwardUnicode(
 
 // Match the input(from the position of start) with relaxed pattern forward.
 // Returns a pair:
-// - first: a bool indicates whether matches the pattern.
-// - second: an integer indicates where is cursor in the input when we finished
-// matching if 'first' is true, -1 otherwise.
+//   - first: a bool which tells us whether it matches.
+//   - second: an integer which has different meaning depends on 'first'.
+//     - where is cursor in the input when we finished matching if 'first' is
+//     true.
+//     - the sub pattern index if 'first' is false, -1 if it fails not because
+//     of any sub-pattern(e.g. the input is too short).
 template <bool isAscii>
 std::pair<bool, int32_t> matchRelaxedFixedForward(
     StringView input,
     const PatternMetadata& patternMetadata,
-    size_t start) {
+    size_t start = 0,
+    size_t startSubPatternIndex = 0) {
   if constexpr (isAscii) {
-    return matchRelaxedFixedForwardAscii(input, patternMetadata, start);
+    return matchRelaxedFixedForwardAscii(
+        input,
+        patternMetadata,
+        start,
+        startSubPatternIndex,
+        patternMetadata.numSubPatterns() - 1);
```

Comment:
Do we always call matchRelaxedFixedForwardAscii and matchRelaxedFixedForwardUnicode with endSubPatternIndex equal to patternMetadata.numSubPatterns() - 1 ? If so, let's remove this argument to make the signature and calling code simpler.

Reply:
Still wondering about this.